Sources/Gate/Error.swift

```diff
@@ -1,11 +1,17 @@
 public struct Unauthorized: Error {
-    let user: Any?
-    let ability: Any
-    let object: Any?
+    public let user: Any
+    public let ability: Any
+    public let object: Any
     
     init<User,Object,Ability:AbilitySet>(user: User?, ability: Ability, object: Object?) {
-        self.user = user
+        self.user = user ?? User.self
```

Comment:
An example of `Any` being used here in an unpractical way. So the `user` is an instance of the type `User`. But you fall back to `User.self` instead which is not an instance but a meta type.

If someone takes the public API when catching/using this error the results are likely unexpected until they dive into the code. One thing you could wonder is if this error actually benefits from exposing the properties publicly. What is the goal this error is meant to achieve? Just an error that is thrown and can be debugged by logging/printing it? Or are you expecing the receiving end to catch the `Unauthorized` error and use the values `user`, `ability` and `object` for other purposes?

In the second scenario, is there any reason the unauthorized struct isn't generic rather than the initializer?

```swift
struct Unauthorized<User, Object, Ability: AbilitySet> {
  let user: User?
  let ability: Ability
  let object: Object?
}
```

Reply:
Okay so here are my reasons for having designed this code like this:

I was imagining that this library might be used by people who are building something with the [vapor framework](https://github.com/vapor/vapor). And then they might want to catch the error quite high up in their app. Which means that at that point they might know an `Unauthorized` error could be thrown, but they'd have no idea which user or object that `Unauthorized` error might be about. So if I make the struct generic, then they would have no way of specifically catching that error if they don't know what user and object it's about.

But now that I think about it, I could make the struct generic and then make it conform to a type-erased protocol so that I kind of have both.

And the reason I did `self.user = user ?? User.self` is because if the `user` is `nil`, then I _at least_ want to be able to save what type the user would've been, had it been filled.

But anyway, I'm gonna try to get best of both worlds with a generic struct conforming to a type-erased protocol.

Reply:
That's a good idea reasoning. I'll retract my suggestion 👍 I know the vapor internals extremely well (having been a [core contributor](https://github.com/vapor/vapor/graphs/contributors). There are a lot of gems not exposed in the docs or not exposed well enough. I could help you shine light on them if you want.